design/metadata-handling.md

```diff
@@ -0,0 +1,286 @@
+  Metal3Machine should result in the same behaviour as before this proposal is
+  implemented
+- the metadata template field would be a list of objects giving the key of the
+  metadata and information on how to retrieve the value, such as an object
+  reference to the metadata object and the key in that object to use.
+- Documentation of this feature would be very important.
+- The controllers must be able to recreate all status objects, i.e. no
+  necessary information should be stored in the status without being recoverable
+
+### Risks and Mitigations
+
+From a user point of view, if this is not well documented or understood, it
+could lead to unexpected errors or behaviours. The errors should be easily
+understandable.
+
+## Design Details
+
+BareMetalHost would have an added field `metaData` in the `Spec` pointing to a
+secret containing the metadata map. Bare Metal Operator would then pass this map
+with added default key/value pairs if not present to Ironic.
+
+A new object would be created, a Metal3Metadata type.
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Metadata
+metadata:
+  name: nodepool-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    controller: true
+    kind: Metal3Cluster
+    name: cluster-1
+spec:
+  values:
+    node-0:
+      abc: def
+    node-1:
+      abc: ghi
+  templates:
+    local-hostname: worker-np1-{{ getIndex }}
+```
+
+This object will be reconciled by the Metal3machine controller. When reconciled,
+the controller will add an OwnerReference to the Metal3Cluster that has nodes
+linking to this object. The spec contains two fields:
+
+- `values` contains a map of id and map of key and values. The id is merely an
+  identifier of the map of key and values.
+- `templates` contains a map of key and values that will be reused for all.
+
+In both cases, the values will be [go templates](
+https://golang.org/pkg/text/template/). A function `getIndex` will return
+the lower int that is not yet in use.
+
+A second object
+would be created, a Metal3MetadataStatus type, linked to a Metal3Cluster.
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3MetadataStatus
+metadata:
+  name: nodepool-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    controller: true
+    kind: Metal3Metadata
+    name: nodepool-1
+spec:
+  values:
+    "node-0": default/machine-1
+  indexes:
+    "0": default/machine-1
+  machines:
+    "default/machine-1":
+      index: "0"
+      value: "node-0"
+```
+
+This object would store the list of values used for each Metal3Metadata object.
+The rationale behind not including it as a status of the previous object is to
+be able to go through the move phase without information loss. This object would
+only be used by the controller and should not be edited by the user.
```

Comment:
Placing status values in the spec of a different CRD is an anti-pattern. It subverts the idea of "spec" being input and "status" being output in the API. It also leads to race conditions (it's impossible to know whether the "indexes" are correctly up to date, because we can't tell which instance of the CRD to look at and whether it has been reconciled by its controller (that's why random suffixes are used everywhere else in kubernetes).

Reply:
Unfortunately random suffixes do not work for our use case. However, I'll make the changes, so that we do not use this object as a database. I was not very happy with this either, but I did not see a good alternative. You proposed one in your following comments that would solve this problem.